File: squint/js/rosetta-squint/tests/integration.test.ts
import { describe, it, expect } from "vitest";
import { readFile } from "node:fs/promises";
import { join } from "node:path";

import {
    phash, phashBytes,
    averageHash, averageHashBytes,
    dhash, dhashBytes,
    dhashVertical, dhashVerticalBytes,
    phashSimple, phashSimpleBytes,
    whashHaar, whashHaarBytes,
    whashDb4, whashDb4Bytes,
    colorhash, colorhashBytes,
    cropResistantHash, cropResistantHashBytes,
    decodeFile,
    hexToHash, hexToFlathash, hexToMultiHash,
    ImageMultiHash,
} from "../src/index.js";
import * as rih from "rosetta-image-hash";

const FIXTURES_PNG = join(
    new URL(".", import.meta.url).pathname,
    "../../../../decode/spec/fixtures/png/valid",
);
const FIXTURES_JPEG = join(
    new URL(".", import.meta.url).pathname,
    "../../../../decode/spec/fixtures/jpeg/valid",
);

const PNG_FILES = [
    "imagehash.png",
    "peppers.png",
    "checker-256.png",
];
const JPEG_FILES = [
    "larger-photo-128.jpg",
    "64x64-quality-50.jpg",
];

// ---------------------------------------------------------------------------
// phash
// ---------------------------------------------------------------------------
describe("phash", () => {
    for (const name of PNG_FILES) {
        const path = join(FIXTURES_PNG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await phash(path, 8);
            expect(h).toBeInstanceOf(rih.Hash);
            expect(h.toHex().length).toBe(16); // 8*8 bits = 64 bits = 16 hex chars
        });

        it(`phash === phashBytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                phash(path, 8),
                readFile(path).then(b => phashBytes(new Uint8Array(b), 8)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });

        it(`phash matches rih.phash(decodeFile) for ${name}`, async () => {
            const [hSquint, img] = await Promise.all([
                phash(path, 8),
                decodeFile(path),
            ]);
            const hRih = rih.phash(img, 8);
            expect(hSquint.toHex()).toBe(hRih.toHex());
        });
    }

    for (const name of JPEG_FILES) {
        const path = join(FIXTURES_JPEG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await phash(path, 8);
            expect(h).toBeInstanceOf(rih.Hash);
        });

        it(`phash === phashBytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                phash(path, 8),
                readFile(path).then(b => phashBytes(new Uint8Array(b), 8)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });
    }
});

// ---------------------------------------------------------------------------
// averageHash
// ---------------------------------------------------------------------------
describe("averageHash", () => {
    for (const name of PNG_FILES) {
        const path = join(FIXTURES_PNG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await averageHash(path, 8);
            expect(h).toBeInstanceOf(rih.Hash);
        });

        it(`averageHash === averageHashBytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                averageHash(path, 8),
                readFile(path).then(b => averageHashBytes(new Uint8Array(b), 8)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });

        it(`averageHash matches rih.averageHash(decodeFile) for ${name}`, async () => {
            const [hSquint, img] = await Promise.all([
                averageHash(path, 8),
                decodeFile(path),
            ]);
            const hRih = rih.averageHash(img, 8);
            expect(hSquint.toHex()).toBe(hRih.toHex());
        });
    }
});

// ---------------------------------------------------------------------------
// dhash
// ---------------------------------------------------------------------------
describe("dhash", () => {
    for (const name of ["imagehash.png", "checker-256.png"]) {
        const path = join(FIXTURES_PNG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await dhash(path, 8);
            expect(h).toBeInstanceOf(rih.Hash);
        });

        it(`dhash === dhashBytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                dhash(path, 8),
                readFile(path).then(b => dhashBytes(new Uint8Array(b), 8)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });
    }
});

// ---------------------------------------------------------------------------
// dhashVertical
// ---------------------------------------------------------------------------
describe("dhashVertical", () => {
    for (const name of ["imagehash.png", "checker-256.png"]) {
        const path = join(FIXTURES_PNG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await dhashVertical(path, 8);
            expect(h).toBeInstanceOf(rih.Hash);
        });

        it(`dhashVertical === dhashVerticalBytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                dhashVertical(path, 8),
                readFile(path).then(b => dhashVerticalBytes(new Uint8Array(b), 8)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });
    }
});

// ---------------------------------------------------------------------------
// phashSimple
// ---------------------------------------------------------------------------
describe("phashSimple", () => {
    for (const name of ["imagehash.png", "peppers.png"]) {
        const path = join(FIXTURES_PNG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await phashSimple(path, 8);
            expect(h).toBeInstanceOf(rih.Hash);
        });

        it(`phashSimple === phashSimpleBytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                phashSimple(path, 8),
                readFile(path).then(b => phashSimpleBytes(new Uint8Array(b), 8)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });
    }
});

// ---------------------------------------------------------------------------
// whashHaar
// ---------------------------------------------------------------------------
describe("whashHaar", () => {
    for (const name of ["imagehash.png", "peppers.png"]) {
        const path = join(FIXTURES_PNG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await whashHaar(path, 8);
            expect(h).toBeInstanceOf(rih.Hash);
        });

        it(`whashHaar === whashHaarBytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                whashHaar(path, 8),
                readFile(path).then(b => whashHaarBytes(new Uint8Array(b), 8)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });
    }
});

// ---------------------------------------------------------------------------
// whashDb4
// ---------------------------------------------------------------------------
describe("whashDb4", () => {
    for (const name of ["imagehash.png", "peppers.png"]) {
        const path = join(FIXTURES_PNG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await whashDb4(path, 8);
            expect(h).toBeInstanceOf(rih.Hash);
        });

        it(`whashDb4 === whashDb4Bytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                whashDb4(path, 8),
                readFile(path).then(b => whashDb4Bytes(new Uint8Array(b), 8)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });
    }
});

// ---------------------------------------------------------------------------
// colorhash
// ---------------------------------------------------------------------------
describe("colorhash", () => {
    for (const name of ["imagehash.png", "peppers.png"]) {
        const path = join(FIXTURES_PNG, name);

        it(`returns a Hash for ${name}`, async () => {
            const h = await colorhash(path, 3);
            expect(h).toBeInstanceOf(rih.Hash);
        });

        it(`colorhash === colorhashBytes for ${name}`, async () => {
            const [hPath, hBytes] = await Promise.all([
                colorhash(path, 3),
                readFile(path).then(b => colorhashBytes(new Uint8Array(b), 3)),
            ]);
            expect(hPath.toHex()).toBe(hBytes.toHex());
        });

        it(`colorhash matches rih.colorhash(decodeFile) for ${name}`, async () => {
            const [hSquint, img] = await Promise.all([
                colorhash(path, 3),
                decodeFile(path),
            ]);
            const hRih = rih.colorhash(img, 3);
            expect(hSquint.toHex()).toBe(hRih.toHex());
        });
    }
});

// ---------------------------------------------------------------------------
// cropResistantHash
// ---------------------------------------------------------------------------
describe("cropResistantHash", () => {
    for (const name of ["imagehash.png", "peppers.png"]) {
        const path = join(FIXTURES_PNG, name);

        it(`returns an ImageMultiHash for ${name}`, async () => {
            const mh = await cropResistantHash(path);
            expect(mh).toBeInstanceOf(ImageMultiHash);
            expect(mh.segmentHashes.length).toBeGreaterThan(0);
        });

        it(`cropResistantHash === cropResistantHashBytes for ${name}`, async () => {
            const [mhPath, mhBytes] = await Promise.all([
                cropResistantHash(path),
                readFile(path).then(b => cropResistantHashBytes(new Uint8Array(b))),
            ]);
            expect(mhPath.toString()).toBe(mhBytes.toString());
        });

        it(`cropResistantHash matches rih.cropResistantHash(decodeFile) for ${name}`, async () => {
            const [mhSquint, img] = await Promise.all([
                cropResistantHash(path),
                decodeFile(path),
            ]);
            const mhRih = rih.cropResistantHash(img);
            expect(mhSquint.toString()).toBe(mhRih.toString());
        });
    }
});

// ---------------------------------------------------------------------------
// Re-export sanity checks
// ---------------------------------------------------------------------------
describe("re-exports", () => {
    it("hexToHash round-trips a 16-char hex string", () => {
        const hex = "ffffffffffffffff";
        const h = hexToHash(hex);
        expect(h.toHex()).toBe(hex);
    });

    it("hexToFlathash round-trips a colorhash string", async () => {
        const path = join(FIXTURES_PNG, "imagehash.png");
        const h = await colorhash(path, 3);
        const hex = h.toHex();
        const h2 = hexToFlathash(hex, 3);
        expect(h2.toHex()).toBe(hex);
    });

    it("hexToMultiHash round-trips a cropResistantHash string", async () => {
        const path = join(FIXTURES_PNG, "imagehash.png");
        const mh = await cropResistantHash(path);
        const str = mh.toString();
        const mh2 = hexToMultiHash(str);
        expect(mh2.toString()).toBe(str);
    });

    it("ImageMultiHash is exported and constructible", () => {
        const bits = [[true, false], [false, true]];
        const h = new rih.Hash(bits);
        const mh = new ImageMultiHash([h]);
        expect(mh.segmentHashes.length).toBe(1);
    });
});

File: squint/rust/rosetta-squint/tests/integration_test.rs
//! Integration tests for rosetta-squint.
//!
//! For each algorithm:
//!   1. `<algo>(path)` returns a non-empty Hash.
//!   2. `<algo>(path)` and `<algo>_bytes(read(path))` produce identical hex.
//!   3. Both are equivalent to calling rosetta_image_hash directly on the decoded image
//!      (proves the decode → hash chain is self-consistent).

use std::fs;
use std::path::Path;

use rosetta_squint::{decode_to_image, *};

// Fixture paths relative to workspace root
const PNG_FIXTURE: &str =
    "../../../decode/spec/fixtures/png/valid/checker-32.png";
const JPEG_FIXTURE: &str =
    "../../../decode/spec/fixtures/jpeg/valid/32x32-quality-95.jpg";
const WEBP_FIXTURE: &str =
    "../../../decode/spec/fixtures/webp/valid/32x32-lossless.webp";

// Large enough for crop_resistant_hash (needs ≥500-pixel segments at 300×300)
const PNG_LARGE_FIXTURE: &str =
    "../../../decode/spec/fixtures/png/valid/checker-256.png";

fn abs(rel: &str) -> std::path::PathBuf {
    let manifest = Path::new(env!("CARGO_MANIFEST_DIR"));
    manifest.join(rel)
}

fn read_fixture(rel: &str) -> Vec<u8> {
    fs::read(abs(rel)).unwrap_or_else(|e| panic!("failed to read {}: {}", rel, e))
}

// ── helpers to get a DynamicImage by going through the same chain ────────────

fn img_from_bytes(bytes: &[u8]) -> image::DynamicImage {
    decode_to_image(bytes).expect("decode_to_image failed")
}

// ── phash ────────────────────────────────────────────────────────────────────

#[test]
fn phash_png_path_and_bytes_agree() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = phash(&path, 8).expect("phash path failed");
    let from_bytes = phash_bytes(&bytes, 8).expect("phash bytes failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), from_bytes.to_hex());
}

#[test]
fn phash_jpeg_chain_consistent() {
    let path = abs(JPEG_FIXTURE);
    let bytes = read_fixture(JPEG_FIXTURE);

    let from_path = phash(&path, 8).expect("phash path failed");
    let from_bytes = phash_bytes(&bytes, 8).expect("phash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::phash(&img, 8).expect("direct phash failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

#[test]
fn phash_webp_chain_consistent() {
    let path = abs(WEBP_FIXTURE);
    let bytes = read_fixture(WEBP_FIXTURE);

    let from_path = phash(&path, 8).expect("phash path failed");
    let from_bytes = phash_bytes(&bytes, 8).expect("phash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::phash(&img, 8).expect("direct phash failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── phash_simple ─────────────────────────────────────────────────────────────

#[test]
fn phash_simple_png_chain_consistent() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = phash_simple(&path, 8).expect("phash_simple path failed");
    let from_bytes = phash_simple_bytes(&bytes, 8).expect("phash_simple bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::phash_simple(&img, 8).expect("direct phash_simple failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

#[test]
fn phash_simple_jpeg_chain_consistent() {
    let path = abs(JPEG_FIXTURE);
    let bytes = read_fixture(JPEG_FIXTURE);

    let from_path = phash_simple(&path, 8).expect("phash_simple path failed");
    let from_bytes = phash_simple_bytes(&bytes, 8).expect("phash_simple bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::phash_simple(&img, 8).expect("direct phash_simple failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── dhash ────────────────────────────────────────────────────────────────────

#[test]
fn dhash_png_chain_consistent() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = dhash(&path, 8).expect("dhash path failed");
    let from_bytes = dhash_bytes(&bytes, 8).expect("dhash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::dhash(&img, 8).expect("direct dhash failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

#[test]
fn dhash_jpeg_chain_consistent() {
    let path = abs(JPEG_FIXTURE);
    let bytes = read_fixture(JPEG_FIXTURE);

    let from_path = dhash(&path, 8).expect("dhash path failed");
    let from_bytes = dhash_bytes(&bytes, 8).expect("dhash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::dhash(&img, 8).expect("direct dhash failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

#[test]
fn dhash_webp_chain_consistent() {
    let path = abs(WEBP_FIXTURE);
    let bytes = read_fixture(WEBP_FIXTURE);

    let from_path = dhash(&path, 8).expect("dhash path failed");
    let from_bytes = dhash_bytes(&bytes, 8).expect("dhash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::dhash(&img, 8).expect("direct dhash failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── dhash_vertical ───────────────────────────────────────────────────────────

#[test]
fn dhash_vertical_png_chain_consistent() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = dhash_vertical(&path, 8).expect("dhash_vertical path failed");
    let from_bytes = dhash_vertical_bytes(&bytes, 8).expect("dhash_vertical bytes failed");
    let img = img_from_bytes(&bytes);
    let direct =
        rosetta_image_hash::dhash_vertical(&img, 8).expect("direct dhash_vertical failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── average_hash ─────────────────────────────────────────────────────────────

#[test]
fn average_hash_png_chain_consistent() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = average_hash(&path, 8).expect("average_hash path failed");
    let from_bytes = average_hash_bytes(&bytes, 8).expect("average_hash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::average_hash(&img, 8).expect("direct average_hash failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

#[test]
fn average_hash_jpeg_chain_consistent() {
    let path = abs(JPEG_FIXTURE);
    let bytes = read_fixture(JPEG_FIXTURE);

    let from_path = average_hash(&path, 8).expect("average_hash path failed");
    let from_bytes = average_hash_bytes(&bytes, 8).expect("average_hash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::average_hash(&img, 8).expect("direct average_hash failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── whash_haar ───────────────────────────────────────────────────────────────

#[test]
fn whash_haar_png_chain_consistent() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = whash_haar(&path, 8).expect("whash_haar path failed");
    let from_bytes = whash_haar_bytes(&bytes, 8).expect("whash_haar bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::whash_haar(&img, 8).expect("direct whash_haar failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

#[test]
fn whash_haar_jpeg_chain_consistent() {
    let path = abs(JPEG_FIXTURE);
    let bytes = read_fixture(JPEG_FIXTURE);

    let from_path = whash_haar(&path, 8).expect("whash_haar path failed");
    let from_bytes = whash_haar_bytes(&bytes, 8).expect("whash_haar bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::whash_haar(&img, 8).expect("direct whash_haar failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── whash_db4 ────────────────────────────────────────────────────────────────

#[test]
fn whash_db4_png_chain_consistent() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = whash_db4(&path, 8).expect("whash_db4 path failed");
    let from_bytes = whash_db4_bytes(&bytes, 8).expect("whash_db4 bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::whash_db4(&img, 8).expect("direct whash_db4 failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

#[test]
fn whash_db4_webp_chain_consistent() {
    let path = abs(WEBP_FIXTURE);
    let bytes = read_fixture(WEBP_FIXTURE);

    let from_path = whash_db4(&path, 8).expect("whash_db4 path failed");
    let from_bytes = whash_db4_bytes(&bytes, 8).expect("whash_db4 bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::whash_db4(&img, 8).expect("direct whash_db4 failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── whash_db4_robust ─────────────────────────────────────────────────────────

#[test]
fn whash_db4_robust_png_chain_consistent() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = whash_db4_robust(&path, 8).expect("whash_db4_robust path failed");
    let from_bytes = whash_db4_robust_bytes(&bytes, 8).expect("whash_db4_robust bytes failed");
    let img = img_from_bytes(&bytes);
    let direct =
        rosetta_image_hash::whash_db4_robust(&img, 8).expect("direct whash_db4_robust failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── colorhash ────────────────────────────────────────────────────────────────

#[test]
fn colorhash_png_chain_consistent() {
    let path = abs(PNG_FIXTURE);
    let bytes = read_fixture(PNG_FIXTURE);

    let from_path = colorhash(&path, 3).expect("colorhash path failed");
    let from_bytes = colorhash_bytes(&bytes, 3).expect("colorhash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::colorhash(&img, 3).expect("direct colorhash failed");

    assert!(!from_path.to_hex().is_empty());
    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

#[test]
fn colorhash_jpeg_chain_consistent() {
    let path = abs(JPEG_FIXTURE);
    let bytes = read_fixture(JPEG_FIXTURE);

    let from_path = colorhash(&path, 3).expect("colorhash path failed");
    let from_bytes = colorhash_bytes(&bytes, 3).expect("colorhash bytes failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::colorhash(&img, 3).expect("direct colorhash failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
    assert_eq!(from_bytes.to_hex(), direct.to_hex());
}

// ── crop_resistant_hash ───────────────────────────────────────────────────────

#[test]
fn crop_resistant_hash_png_path_and_bytes_agree() {
    let path = abs(PNG_LARGE_FIXTURE);
    let bytes = read_fixture(PNG_LARGE_FIXTURE);

    let from_path = crop_resistant_hash(&path, None).expect("crop_resistant_hash path failed");
    let from_bytes =
        crop_resistant_hash_bytes(&bytes, None).expect("crop_resistant_hash bytes failed");

    assert!(!from_path.segment_hashes.is_empty());
    assert_eq!(from_path.to_hex(), from_bytes.to_hex());
}

#[test]
fn crop_resistant_hash_png_chain_consistent() {
    let path = abs(PNG_LARGE_FIXTURE);
    let bytes = read_fixture(PNG_LARGE_FIXTURE);

    let from_path = crop_resistant_hash(&path, None).expect("crop_resistant_hash path failed");
    let img = img_from_bytes(&bytes);
    let direct = rosetta_image_hash::crop_resistant_hash(&img, None)
        .expect("direct crop_resistant_hash failed");

    assert_eq!(from_path.to_hex(), direct.to_hex());
}

#[test]
fn crop_resistant_hash_limit_segments_caps_count() {
    // Verifies the limit_segments parameter is forwarded end-to-end (squint → hash).
    let path = abs(PNG_LARGE_FIXTURE);
    let unlimited =
        crop_resistant_hash(&path, None).expect("crop_resistant_hash unlimited failed");
    if unlimited.segment_hashes.len() <= 1 {
        // The chosen fixture only produces one segment; nothing to verify.
        return;
    }
    let limited =
        crop_resistant_hash(&path, Some(1)).expect("crop_resistant_hash limit=1 failed");
    assert_eq!(limited.segment_hashes.len(), 1);
    assert!(limited.segment_hashes.len() < unlimited.segment_hashes.len());
}

// ── print example hash for report ────────────────────────────────────────────

#[test]
fn report_phash_on_grayscale_jpeg() {
    // This test exists solely to emit the example hash for the report.
    let grayscale_jpg =
        "../../../decode/spec/fixtures/jpeg/valid/8x8-grayscale.jpg";
    let path = abs(grayscale_jpg);
    if !path.exists() {
        return; // fixture missing, skip
    }
    let h = phash(&path, 8).expect("phash on 8x8-grayscale.jpg failed");
    // Print the hash so it appears in `cargo test -- --nocapture` output.
    println!("phash(8x8-grayscale.jpg, 8) = {}", h);
    assert!(!h.to_hex().is_empty());
}
